### DISCUSSION TITLE
rtlsdr_write_reg failed with -1 -4 -7

### DISCUSSION THREAD

**Discussion by kenwawa01** [merbanan/rtl_433#2968] [Q&A]
Hi!

I've lately changed the device being used with rtl_433 to this one: https://pl.aliexpress.com/item/32965619493.html

I use ubuntu 22.04 - tried few different kernels, the newest 6.5.0-35.

I expected that there might be some problems but... after long fight I'm not able for force it to any work. 

The most common problem reported is:

```
Found Rafael Micro R828D tuner
r82xx_write: i2c wr failed=-7 reg=0c len=1
r82xx_init: failed=-7
rtlsdr_demod_write_reg failed with -9
rtlsdr_demod_read_reg failed with -1
[SDR] Using device 0: Realtek, RTL2838UHIDIR, SN: 00000001, "Generic RTL2832U OEM"
Exact sample rate is: 250000.000414 Hz
rtlsdr_demod_write_reg failed with -9
r82xx_write: i2c wr failed=-1 reg=0a len=1

```
But when I use 

> rtl_test

 program I get for example:

```
Found 1 device(s):
  0:  , , SN: 

Using device 0: Generic RTL2832U OEM
rtlsdr_write_reg failed with -1
Resetting device...
rtlsdr_write_reg failed with -4
rtlsdr_write_reg failed with -4
rtlsdr_write_reg failed with -4
rtlsdr_write_reg failed with -4
rtlsdr_write_reg failed with -4
rtlsdr_demod_write_reg failed with -4
rtlsdr_demod_read_reg failed with -4
rtlsdr_demod_write_reg failed with -4
rtlsdr_demod_read_reg failed with -4
```

Another example (in different usb port but that probably doesnt count and its rather random):

Found 1 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: 00000001

Using device 0: Generic RTL2832U OEM
Found Rafael Micro R828D tuner
r82xx_write: i2c wr failed=-1 reg=1a len=7
r82xx_write: i2c wr failed=-1 reg=0c len=1
r82xx_init: failed=-1
rtlsdr_demod_write_reg failed with -1
rtlsdr_demod_read_reg failed with -1
Supported gain values (29): 0.0 0.9 1.4 2.7 3.7 7.7 8.7 12.5 14.4 15.7 16.6 19.7 20.7 22.9 25.4 28.0 29.7 32.8 33.8 36.4 37.2 38.6 40.2 42.1 43.4 43.9 44.5 48.0 49.6 
rtlsdr_demod_write_reg failed with -1
rtlsdr_demod_read_reg failed with -1
r82xx_write: i2c wr failed=-1 reg=0a len=1

I tried with few different version of kernels. I updated libusb1.0 and rtlsdr libraries.

Any idea how to force it to work with with this device? Thanks for help and suggestions.

**Comment by zuckschwerdt** [comment]
Did you try some other USB cable? Also try to ensure that the USB power is stable.

**Comment by kenwawa01** [comment]
Ok I solved it by myself.

First I didn't clean the old version of librtlsdrlibrary. Secondly I used this library https://github.com/steve-m/librtlsdr 2.0.2.

When cleaning, Installing the newest master version of this https://github.com/librtlsdr/librtlsdr and updating kernel to 6.5.0-35, everything works flawlessly. Not sure if all 3 steps were necessary. Sure I also recompiled rtl_433.

Btw, the given device at the beginning has very nice sensitivity on 433MHz. It's similar price as dvb-t/t2 tuners but the signal is much better! It might be also related with better antenna attached by default (here I'm not so sure how much its antenna effect and how much better design of this device counts...).

**Comment by kenwawa01** [comment]
However it looks like that after writing that everything works ok, I discovered that... it's not always the case and it sometimes happens that I'm getting error codes but much more seldom. If it starts working, it works without problems. But sometimes I have to "kick the engine" few times...

Any ideas by what it might be caused?

I tried to monitor the situation with 

`sudo udevadm monitor`

and it's one of the output (the device was unbind despite I didn't disconnect it, at first glance it looks like a problem with cables, stable connection etc. but when moving cables or device itself I can't see any improvements when it's not working or it's not going worse when it's working ok thus for me it looks rather some software error):

KERNEL[22295.009335] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
KERNEL[22295.014909] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.0 (usb)
KERNEL[22295.016782] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.1 (usb)
KERNEL[22295.016799] bind     /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
UDEV  [22295.032982] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
UDEV  [22295.036721] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.0 (usb)
UDEV  [22295.037504] add      /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.1 (usb)
UDEV  [22295.043067] bind     /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
KERNEL[22324.376215] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.0 (usb)
KERNEL[22324.376243] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.1 (usb)
KERNEL[22324.376355] unbind   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
KERNEL[22324.376371] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
UDEV  [22324.380267] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.0 (usb)
UDEV  [22324.380509] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1/5-1:1.1 (usb)
UDEV  [22324.383620] unbind   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
UDEV  [22324.386685] remove   /devices/pci0000:00/0000:00:08.1/0000:07:00.4/usb5/5-1 (usb)
